[DONT MERGE YET][dynamo][guards] Do not dict pop in framelocals dict creation

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164038

For this program

```
import torch

torch._dynamo.config.enable_cpp_symbolic_shape_guards = False

def fn(x):
    for _ in range(20000):
        x = x + 1
    return x

mod = torch.fx.symbolic_trace(fn)

print(mod)


opt_mod = torch.compile(mod, backend="eager", dynamic=True)

x = torch.randn(2, 2)
opt_mod(x)

```

It reduces the guard overhead from 1.8 ms to 72 us. The dict pop is not required as explained in the comment.

It can reduce to 46 us if we just remove the _dict.attr pop branch.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 committed:
[dynamo][guards] Do not dict pop in framelocals dict creation

[ghstack-poisoned]

mlazos reviewed (approved):
really cool speedup!

anijain2305 committed:
Update on "[dynamo][guards] Do not dict pop in framelocals dict creation"


For this program

```
import torch

torch._dynamo.config.enable_cpp_symbolic_shape_guards = False

def fn(x):
    for _ in range(20000):
        x = x + 1
    return x

mod = torch.fx.symbolic_trace(fn)

print(mod)


opt_mod = torch.compile(mod, backend="eager", dynamic=True)

x = torch.randn(2, 2)
opt_mod(x)

```

It reduces the guard overhead from 1.8 ms to 172 us. The dict pop is not required as explained in the comment.

It can reduce to 46 us if we just remove the _dict.attr pop branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Do not dict pop in framelocals dict creation"


For this program

```
import torch

torch._dynamo.config.enable_cpp_symbolic_shape_guards = False

def fn(x):
    for _ in range(20000):
        x = x + 1
    return x

mod = torch.fx.symbolic_trace(fn)

print(mod)


opt_mod = torch.compile(mod, backend="eager", dynamic=True)

x = torch.randn(2, 2)
opt_mod(x)

```

It reduces the guard overhead from 1.8 ms to 72 us. The dict pop is not required as explained in the comment.

It can reduce to 46 us if we just remove the _dict.attr pop branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 commented:
Not merging yet. @williamwen42  has concerns and he will take over this PR.
